베이스볼 코드 리뷰 부탁드립니다.

sonrose opened the pull request:
### 구현 내용

1. **Input Class**
   사용자에게 값을 입력 받는 클래스

   - **input()**
     입력 받은 문자열을 하나씩 분리해 배열 number에 넣는다.
     값에 중복이 있는지, 3글자가 맞는지 검사한 후 이를 만족하지 못하면 재입력 받는다.
     배열에 있는 값을 정수로 변환한 뒤 리스트에 넣고 반환한다.
   
   - **clear()**
     리스트를 초기화한다.

   - **inputData()**
     값을 입력 받고 한 글자씩 분리한다.
   
   - **checkForLength(String[] nums)**
     문자열의 길이가 3인지 검사한다. 만약 3이 아닐 경우 true를 반환한다.

   - **checkForDuplicates(String[] nums)**
     중복값이 있는지 검사한다. 만약 중복값이 있을 경우 true를 반환한다. (배열의 총 개수와 중복값을 걸러낸 개수 비교)

---

2. **RandomNumber Class**
    랜덤한 정수를 생성하는 클래스
    
    - **createRandomNumber()**
      중복값이 없도록 Set으로 값을 저장한다. 리스트에 3개가 들어갈 때까지 반복해서 랜덤값을 생성하고 넣는다.

---

3. **Comparison Class**
    사용자가 입력한 값과 랜덤 생성한 값을 비교하는 클래스

    - **compareNumber(ArrayList<Integer> input, ArrayList<Integer> randomNumber)**
      만약 입력 값과 랜덤값 중 순서 상관없이 겹치는 값이 있다면 ball을 하나 늘린다.
      순서도 같이 겹치는 값이 있다면 strike를 하나 늘린다.

---

4. **Output Class**
    결과를 출력하는 클래스

    - **output(int strike, int ball)**
      strike와 ball 값을 받아와, 만약 strike가 3이면 답을 맞힌 것이므로 BASEBALL WIN~!을 출력한다.
      정답을 맞히지 못했다면 ball, strike 여부를 알려준다.

---

### 궁금한 점

main에 있는 while문... 실행문 안에 아무것도 안 들어가도 되나요...

sonrose committed:
베이스볼 구현 1

DongGeon0908 commented on src/main/java/RandomNumber.java:
```diff
@@ -0,0 +1,18 @@
+
+public class RandomNumber {
+
+    private static final Set<Integer> randomNumbersSet = new HashSet<>();
```
해당 변수를 메모리에서 관리해야 하나요?
왜 static을 쓰신거죠?

DongGeon0908 replied:
Set이라는 명칭을 사용하지 않습니다!

DongGeon0908 commented on src/main/java/RandomNumber.java:
```diff
@@ -0,0 +1,18 @@
+
+    private static final Set<Integer> randomNumbersSet = new HashSet<>();
+
+    public ArrayList<Integer> createRandomNumber() {
```
```suggestion
    public List<Integer> createRandomNumber() {
```

DongGeon0908 commented on src/main/java/Comparison.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.ArrayList;
+
+public class Comparison {
```
야구라는 도메인에서 필요한 객체는 무엇일까요?

DongGeon0908 commented on src/main/java/Comparison.java:
```diff
@@ -0,0 +1,27 @@
+import java.util.ArrayList;
+
+public class Comparison {
+    int ball;
+    int strike;
```
접근제한자는?

DongGeon0908 commented on src/main/java/Comparison.java:
```diff
@@ -0,0 +1,27 @@
+    int strike;
+
+    public int compareNumber(ArrayList<Integer> input, ArrayList<Integer> randomNumber) {
+        ball = 0;
+        strike = 0;
```
???????초기화를 의미하는 걸까요?

DongGeon0908 commented on src/main/java/Comparison.java:
```diff
@@ -0,0 +1,27 @@
+        strike = 0;
+
+        for (int i = 0; i < 3; i++) {
+            if (input.contains(randomNumber.get(i))) {
+                ball++;
+            }
+
+            if (input.get(i).equals(randomNumber.get(i))) {
+                ball = 0;
+                strike++;
+            }
```
if else 쓰지 않고 한번 도전해볼까요?

DongGeon0908 commented on src/main/java/Comparison.java:
```diff
@@ -0,0 +1,27 @@
+            }
+        }
+
+        Output output = new Output();
+        output.output(strike, ball);
```
여기서 출력을 실행할 이유가 있을까요?

DongGeon0908 commented on src/main/java/Main.java:
```diff
@@ -1,5 +1,15 @@
+
 public class Main {
     public static void main(String[] args) {
         System.out.println("Play Baseball Game");
```
이것도 output으로 빼도 좋을듯

DongGeon0908 commented on src/main/java/Input.java:
```diff
@@ -0,0 +1,40 @@
+public class Input {
+    private ArrayList<Integer> numbers = new ArrayList<>();
+
+    public ArrayList<Integer> input() {
+        clear();
+        System.out.println("\n" + "숫자 3개를 입력해 주세요. (예: 123)");
+        String[] number = inputData();
+
+        while (checkForDuplicates(number) == true || checkForLength(number) == true) {
+            System.out.println("\n" + "입력값은 중복이 없어야 하며 3자리여야 합니다. 재입력해 주세요.");
+            number = inputData();
+        }
+
+        for (int i = 0; i < 3; i++) {
+            numbers.add(Integer.parseInt(number[i]));
+        }
+
+        return numbers;
+    }
```
메서드가 길어진다면, 너무 많은 기능을 하나에서 진행한다고 생각하시면 됩니다

DongGeon0908 reviewed (changes requested):
읽기 쉬운, 그런 코드를 짜보시죵!

DongGeon0908 commented on src/main/java/Main.java:
```diff
@@ -1,5 +1,15 @@
+
+        while (compareNumber.compareNumber(input.input(), randomNumber.createRandomNumber()) != 3) {
+
+        }
```
???????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????

sonrose committed:
베이스볼 구현 2

sonrose committed:
베이스볼 구현 2-1
README.md, Comparison.java 접근제한자(private) 수정

DongGeon0908 commented on README.md:
```diff
@@ -24,3 +24,72 @@
 wrjs@naver.com
```
Readme 작성 좋네요

DongGeon0908 commented on src/main/java/RandomNumber.java:
```diff
@@ -0,0 +1,19 @@
+
+public class RandomNumber {
+
+    private Set<Integer> randomNumber = new HashSet<>();
```
클래스를 생성할 때 만들어주는 건 어떨까요?

DongGeon0908 commented on src/main/java/RandomNumber.java:
```diff
@@ -0,0 +1,19 @@
+            randomNumber.add((int) (Math.random() * 9 + 1));
+        }
+
+        List<Integer> randomNumbers = new ArrayList<>(randomNumber);
+        return randomNumbers;
```
```suggestion
        return new ArrayList<>(randomNumber);
```

DongGeon0908 reviewed (changes requested):
메서드 분리 작업에서 노력하신 것 같네요!
고생하셨어요

이제는 객체와 객체가 서로 메세지를 던지며 상호작용하도록 수정하시죠!!

sonrose committed:
베이스볼 구현 3
객체와 객체가 서로 메세지를 던지며 상호작용하도록 구현하라는 게, 리턴값을 메서드의 파라미터로 넣어서 상호작용하도록 하라는 건가요?
조금 이해가 안 됩니다.

DongGeon0908 reviewed (commented):
굳굳

DongGeon0908 commented on src/main/java/RandomNumber.java:
```diff
@@ -0,0 +1,23 @@
+
+    private final Set<Integer> randomNumber;
+
+    RandomNumber() {
+        randomNumber = new HashSet<>();
+    }
```
접근제한자를 default로 한 이유는 무엇일까요?

DongGeon0908 commented on src/main/java/RandomNumber.java:
```diff
@@ -0,0 +1,23 @@
+            randomNumber.add((int) (Math.random() * 9 + 1));
+        }
+
+        System.out.println(randomNumber);
```
출력하는 부분을 따로 빼는 건 어떨까요?

DongGeon0908 commented on src/main/java/Comparison.java:
```diff
@@ -0,0 +1,42 @@
+
+public class Comparison {
+
+    private int ball;
+    private int strike;
```
생성자로 초기화할 수는 없나용?

DongGeon0908 commented on src/main/java/Comparison.java:
```diff
@@ -0,0 +1,42 @@
+        return new int[] {ball, strike};
+    }
+
+    public int checkForBall(List<Integer> randomNumber, List<Integer> userNumber) {
+
+        for(int i = 0; i < 3; i++) {
+            if (randomNumber.contains(userNumber.get(i))) {
+                ball++;
+            }
+        }
+
+        return ball;
+    }
+
+    public int checkForStrike(List<Integer> randomNumber, List<Integer> userNumber) {
+
+        for(int i = 0; i < 3; i++) {
+            if (randomNumber.get(i).equals(userNumber.get(i))) {
+                strike++;
+                ball--;
+            }
+        }
+
+        return strike;
+    }
```
2개가 아에 같은 로직으로 구성되어 있는 것 같아요!
아에 strike인지, ball인지 데이터를 넘겨주면 바로 알려줄 수 있는 별도의 클래스를 만드는 건 어떨까요>

DongGeon0908 commented on src/main/java/Input.java:
```diff
@@ -0,0 +1,37 @@
+
+    public List<Integer> arrToList(String[] str) {
+
+        List<String> number = Arrays.asList(str);
+        List<Integer> numbers = number.stream().map(s -> Integer.parseInt(s)).collect(Collectors.toList());
```
이거 하나의 문장으로 바꿀 수 있을 것 같아여

DongGeon0908 reviewed (changes requested):
객체가 다른 객체에게 메세지를 던진다는 의미가 무엇일까요??
어려운 개념이긴 하지만, 한번 더 고민해서 수정해보세요!!
(거의 끝나가요!)

sonrose committed:
베이스볼 구현 4
while문에서 객체 생성하면 안 될 것 같은데, 만약 그러지 않으면 Ball과 Strike 값이 초기화가 안 돼서 while문 안에 넣었습니다. 이거 어떻게 방법 없나요...

DongGeon0908 commented on src/main/java/Strike.java:
```diff
@@ -0,0 +1,29 @@
+
+public class Strike extends Baseball {
+
+    private int ball;
```
strike class가 ball도 알고 있어야 할까요?

DongGeon0908 commented on src/main/java/Strike.java:
```diff
@@ -0,0 +1,29 @@
+    private int ball;
+    private int strike;
+
+    Strike() {
```
접근제한자~

DongGeon0908 commented on src/main/java/Strike.java:
```diff
@@ -0,0 +1,29 @@
+public class Strike extends Baseball {
+
+    private int ball;
+    private int strike;
```
final 사용하지 않은 이유가 무엇일까용?
불변객체에 대해 이번에 학습해보시길 추천드립니다

DongGeon0908 commented on src/main/java/Main.java:
```diff
@@ -1,5 +1,29 @@
+            int totalStrike = strike.checkBallOrStrike(randomNumbers, userNumbers);
+            int totalBall = strike.getBall() + ball.checkBallOrStrike(randomNumbers, userNumbers);
+
+            result = new int[]{totalBall, totalStrike};
```
이부분이 이해가 되지 않네요
이미 위에서 선언했는데, 또 생성한 이유가 있나요?

DongGeon0908 commented on src/main/java/Input.java:
```diff
@@ -0,0 +1,36 @@
+        return str.split("");
+    }
+
+    public List<Integer> arrToList(String[] str) {
```
자료구조 명칭은 메서드 네이밍으로 정하지 않습니다

DongGeon0908 commented on src/main/java/Baseball.java:
```diff
@@ -0,0 +1,6 @@
+
+public abstract class Baseball {
+
+    public abstract int checkBallOrStrike(List<Integer> randomNumber, List<Integer> userNumber);
```
해당 추상메서드가 어색한 것 같아요
차라리 check라고 가져가는게 어떨까요?

DongGeon0908 commented on src/main/java/Ball.java:
```diff
@@ -0,0 +1,22 @@
+
+    private int ball;
+
+    Ball() {
```
접근제한자는?

DongGeon0908 reviewed (changes requested):
추상클래스를 쓰신건 좋은데, 조금 더 광범위하게 쓰일 수 있도록 네이밍을 정해주세요!
각 클래스마다 역할과 책임을 다할 수 있도록 분리했으면 좋겠습니다
그리고 불변객체에 대해 공부해보세요

sonrose committed:
베이스볼 구현 5

DongGeon0908 commented on src/main/java/Ball.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    @Override
+    public int check(List<Integer> randomNumber, List<Integer> userNumber) {
```
보통 유효성 검사 네이밍은 check 또는 validate를 사용합니다

DongGeon0908 commented on src/main/java/Ball.java:
```diff
@@ -0,0 +1,24 @@
+    @Override
+    public int check(List<Integer> randomNumber, List<Integer> userNumber) {
+
+        int ball = 0;
+
+        for (int i = 0; i < 3; i++) {
+            if (randomNumber.contains(userNumber.get(i))) {
+                ball++;
+            }
+        }
+
+        return ball;
```
stream으로 한문장안에 끝낼 수 있을 것 같아용

DongGeon0908 commented on src/main/java/Baseball.java:
```diff
@@ -0,0 +1,6 @@
+import java.util.List;
+
+public abstract class Baseball {
```
인터페이스로 만드는 건 어떤까요?

DongGeon0908 commented on src/main/java/Input.java:
```diff
@@ -0,0 +1,36 @@
+    }
+
+    public void validateForDuplication(List<Integer> numbers) {
+        if (numbers.size() != numbers.stream().distinct().count()) {
```
```suggestion
int count =  numbers.stream().distinct().count()

        if (numbers.size() != count) {
```

DongGeon0908 replied:
보통은 if문 안에 조건이 긍정이 되도록 구성하는게 좋아요
읽기 편하거든요

DongGeon0908 commented on src/main/java/Main.java:
```diff
@@ -1,5 +1,28 @@
+        Strike strike = new Strike();
+
+        int totalStrike = 0;
+        int totalBall;
```
초기화하지 않은 값에는 어떤게 들어달까요?

DongGeon0908 commented on src/main/java/Strike.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.List;
+
+public class Strike extends Baseball {
```
Baseball이라는 계층을 주어서 추상화한 이유가 궁금합니다

sonrose replied:
안에 있는 메서드가 중복되어서, 이걸 아예 하나의 baseball 객체(클래스)로 만든 다음에 ball과 strike로 나누면 더 깔끔할 것 같아서 추상 클래스로 만들었습니다.

DongGeon0908 reviewed (changes requested):
고생하셨습니다!!
거의 끝나가네요
몇가지 질문 및 개선사항에 대해 피드백 드렸습니다

DongGeon0908 commented:
한단계 점프업을 위해 배열 -> 리스트로 전부 바꾸시죠!!

sonrose committed:
베이스볼 구현 6

DongGeon0908 commented on src/main/java/Strike.java:
```diff
@@ -0,0 +1,16 @@
+
+    @Override
+    public int compare(List<Integer> randomNumber, List<Integer> userNumber) {
+        return (int) IntStream.range(0, 3).filter(n -> randomNumber.get(n).equals(userNumber.get(n))).count();
```
조금 더 깔끔하게 수정할 수 있을 것 같네요

DongGeon0908 commented on src/main/java/RandomNumber.java:
```diff
@@ -0,0 +1,23 @@
+
+    public List<Integer> createRandomNumber() {
+        while(randomNumber.size() < 3) {
+            randomNumber.add((int) (Math.random() * 9 + 1));
```
정말 랜덤한지 어떻게 증명할 수 있을까요?

sonrose replied:
Math.random이 정말 랜덤한지 찾아보니, 유사 난수를 생성한다고 하더군요. (시드값도 설정 불가)
즉, 난수를 계속 생성하다 보면 어떠한 패턴이 나타난다고 합니다.
따라서 Math.random이 아닌 java.util의 Random을 사용해 시드값을 System.currentTimeMillis()로 설정하였습니다.

그런데 제 생각에는 java.util의 Random도 Math.random보다 좀 더 나을뿐, 결국 유사 난수를 생성하는 건 변함 없는 것 같은데, 제 생각이 맞나요?

DongGeon0908 commented on src/main/java/Main.java:
```diff
@@ -1,5 +1,28 @@
+        Ball ball = new Ball();
+        Strike strike = new Strike();
+
+        int totalStrike = 0;
+        int totalBall = 0;
```
여기에 꼭 필요한 값일까요?

DongGeon0908 commented on src/main/java/Input.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public List<Integer> toNumbers(String str) {
+        List<Integer> numbers = Pattern.compile("").splitAsStream(str).map(Integer::parseInt).collect(Collectors.toList());
```
조금 더 깔끔하게?
코드를 읽기 쉽게 보여주시는게 좋을 것 같네요

sonrose replied:
객체지향 생활체조 원칙 중 <한 줄에 점을 하나만 찍는다.> 이걸 말씀하시는 것 같아 점을 기준으로 나눴어요.
이거 말고는... 읽기 쉽게 한다는 게 어떻게 하는 것인지 잘 모르겠어요.

DongGeon0908 commented on src/main/java/Input.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    public void validateForNumberLength(List<Integer> numbers) {
+        if(numbers.size() > 3 || numbers.size() < 3 || numbers.contains(0)) {
```
해당 부분에 대해 조금 더 나눠 볼 수 있을까요?

DongGeon0908 commented on src/main/java/Ball.java:
```diff
@@ -0,0 +1,16 @@
+
+    @Override
+    public int compare(List<Integer> randomNumber, List<Integer> userNumber) {
+        return (int) IntStream.range(0, 3).filter(n -> randomNumber.contains(userNumber.get(n))).count();
+    }
```
깔끔하게 수정하기!

DongGeon0908 reviewed (commented):
조금 더 읽기 쉬운 코드로 수정하시고, 이번 과제는 끝내도 될 것 같습니다

sonrose committed:
베이스볼 구현 7

DongGeon0908 commented on src/main/java/RandomNumber.java:
```diff
@@ -0,0 +1,24 @@
+        randomNumber = new HashSet<>();
+    }
+
+    public List<Integer> createRandomNumber() {
+        Random random = new Random(System.currentTimeMillis());
+
+        while(randomNumber.size() < 3) {
+            randomNumber.add(random.nextInt(9) + 1);
+        }
+
+        log.info("랜덤값 = {}", randomNumber);
+        return new ArrayList<>(randomNumber);
+    }
```
해당 값이 정말 랜덤한지 validate해주는 로직이 있으면 좋겠네요

DongGeon0908 reviewed (commented):
고생하셨습니다
랜덤 넘버에 대한 유효성 로직만 추가하면 될 것 같아요!!
그 부분만 추가하시고, 다음 단계로 넘어가시죠
pr은 닫을게요~~
